### DISCUSSION TITLE
Moving away from firebase

### DISCUSSION THREAD

**Discussion by lukew3** [monkeytypegame/monkeytype#1405] [Ideas] [answered]
Using firebase will become really expensive as more users join. I'd suggest that the backend be moved from firebase to a linux server with mongodb. Queries should be pretty similar between firebase and mongodb and the cost of running the servers will be much lower and more predictable. 
Also, it would be nice to know how much it costs to run right now if you're comfortable with sharing.

**Comment by Miodec** [comment, marked as answer]
Yeah i know, i want to its just a big change to rewrite all the database connections, move the data over without breaking anything and not loosing data.

Right now its approaching 400, mostly due to the amount of data stored.

**Reply by lukew3** [reply]
I've got a lot of free time now so I'll see if I can migrate the cloud-functions to mongodb. We could get costs way lower, maybe around $100. Check out [this calculator](https://www.linode.com/estimator/)

**Reply by Miodec** [reply]
Yeah that would be pretty great - as mentioned in my other comment, wouldnt this transition make local development a bit more difficult? Or are there easy ways of self hosting a little mongo database? (ideally all without external tools, just with the gulpfile)

**Reply by lukew3** [reply]
The mongo database runs on the same server that sends the frontend, which will probably be powered by node and express.js. If you are serving on localhost, you can install mongo on the computer you are hosting on and connect to it. You just have to [install mongodb community edition](https://docs.mongodb.com/manual/administration/install-community/). You would have to start it outside the gulpfile, but it can start itself on your pc's startup and run in the background if you allow it to. In my opinion, it's easier than firebase, plus it can be worked on completely offline.

I'm working on migrating right now, hoping to get it working by the end of the weekend.

**Reply by Andrew15-5** [reply]
Hey. I looked through repo for related context. I found https://github.com/monkeytypegame/monkeytype/discussions/3961#discussioncomment-5536130 from 2023. This means that you can't move away from firebase, but here it looks like you can. At least if a PR is created. @lukew3, what is the status on migration?

https://github.com/monkeytypegame/monkeytype/blob/master/docs/CONTRIBUTING_ADVANCED.md#firebase-optional:

> The account system will not let you create an account without a Firebase project. You can skip this if you don't think you will need it (you can always set it up later)

I recently was able to move to an offline version of https://github.com/cs0x7f/cstimer, which was actually pretty easy. Just run PHP server and you're good to go. After this, I wanted to move monkeytype offline too, so that I'm not dependent on Internet connection and can store my progress locally, do any sort of manual/automatic backup etc. But Firebase account prevents me from doing this. What is the current roadblock, @Miodec?

I also found https://github.com/monkeytypegame/monkeytype/pull/1475 and https://github.com/monkeytypegame/monkeytype/pull/2312, but only skimmed through. There were some security issues or something?

**Reply by fehmer** [reply]
this discussion is about the usage firebase for monkeytype.com, so the main site. Currently we don't plan to replace firebase.

For self-hosting it would be an option but it would take some effort and has no priority from our side at the moment.

**Reply by Andrew15-5** [reply]
So, technically, it is still possible to see future version with offline self-hosting/local capability? I'm not sure what Luke was cooking, but I would be interested to tackle this problem (at some point in the future). I assume the major issue is just that there are many API endpoints? Considering that the queries will probably be unchanged (NoSQL), the problem is to just get time to write not only a local db version, but also so that you can easily switch between them with env vars. At least, that what I think it does right now. Didn't look into the code.

**Reply by fehmer** [reply]
it is possible, we can discuss in detail when you have time to work on this.

**Reply by Andrew15-5** [reply]
Should this be a separate issue then? I didn't find any open or closed. Maybe the majority are in discussions.

**Reply by fehmer** [reply]
you can start a new discussion if you want. this one is four years old and the project has changed a lot since then

**Reply by Andrew15-5** [reply]
Right, no feature request issues... Here you go: https://github.com/monkeytypegame/monkeytype/discussions/6823.

**Comment by nogweii** [comment]
Would you consider supporting [remoteStorage](https://remotestorage.io/)? It allows people to store their history somewhere else and you wouldn't have to pay for those users.

**Reply by Miodec** [reply]
Whats the difference between remote storage and mongo?

**Reply by lukew3** [reply]
> Whats the difference between remote storage and mongo?

From what I've read, remote storage means that each individual person's data is stored in an outside server (such as  [5apps](https://5apps.com/storage) or self-hosted). The individual has complete control over their data and can then decide how they want their data to be shared between apps and such. The user would have to create an account on another platform where their data is actually stored. With a mongo server, the data is managed by you, the monkeytype database administrator. Using remote storage could save you, the administrator, money since they are not actually hosting data, but would transfer the costs to the individual/database provider.

For now, I think we should design our own backend powered by mongodb and hosted on something like [linode](https://www.linode.com/). Remote storage could be added later if enough users plan to use it to deem it worth the effort.

**Reply by Miodec** [reply]
Yeah i think mongo is the way to go for now. Dunno how it will work with local dev and in general i dont know where to start with the conversion.
